specifications/secure-launch-specification.rst

```diff
@@ -0,0 +1,719 @@
+    :linenos: 1
+
+    struct slr_policy_entry {
+        u16 pcr;
+        u16 entity_type;
+        u16 flags;
+        u16 reserved;
+        u64 entity;
+        u64 size;
+        char evt_info[TPM_EVENT_INFO_LENGTH];
+    };
+
+DRTM Policy Entry Entity Types
+''''''''''''''''''''''''''''''
+
+The list of valid entity types for DRTM Policy entries.
+
+.. code-block:: c
+    :linenos: 1
+
+    #define SLR_ET_UNSPECIFIED  0x0000
+    #define SLR_ET_SLRT         0x0001
+    #define SLR_ET_BOOT_PARAMS  0x0002
+    #define SLR_ET_SETUP_DATA   0x0003
+    #define SLR_ET_CMDLINE      0x0004
+    #define SLR_ET_EFI_MEMMAP   0x0005
+    #define SLR_ET_RAMDISK      0x0006
+    #define SLR_ET_TXT_OS2MLE   0x0010
+    #define SLR_ET_UNUSED       0xffff
+
+DRTM Policy Entry Flags
+'''''''''''''''''''''''
+
+The list of valid flags for DRTM Policy entries.
+
+.. code-block:: c
+    :linenos: 1
+
+    #define SLR_POLICY_FLAG_MEASURED	0x1
+    #define SLR_POLICY_IMPLICIT_SIZE	0x2
+
+Intel TXT Platforms
+~~~~~~~~~~~~~~~~~~~
+
+When on Intel platforms specific information needs to be conveyed to Secure
+Launch.
+
+Intel TXT Info
+^^^^^^^^^^^^^^
+
+:REQUIRED: This entry **MUST** be present on Intel platforms.
+
+Intel TXT requires for the pre-launch environment to pass MSR and MTRR state
+across to the post-launch environment.
+
+:tag: SLR_ENTRY_INTEL_INFO
+:saved_misc_enable_msr: Saved MSR values
+:saved_bsp_mtrrs: Saved BSP MTRRs
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_intel_info {
+        struct slr_entry_hdr hdr;
+        u64 saved_misc_enable_msr;
+        struct txt_mtrr_state saved_bsp_mtrrs;
+    };
+
+Saved MTRR State
+""""""""""""""""
+
+:code:`struct txt_mtrr_state`
+
+:default_mem_type: The default memory type for regions not covered by an MTRR
+:mtrr_vcnt: Number of variable MTRR pairs in the mtrr_vcnt array
+:mtrr_pair: Array of variable MTRR pairs to restore post launch
+
+:code:`struct txt_mtrr_pair`
+
+:mtrr_physbase: Physical base address for variable MTRR
+:mtrr_physmask: Physical mask for the variable MTRR
+
+.. code-block:: c
+    :linenos: 1
+
+    struct txt_mtrr_pair {
+        u64 mtrr_physbase;
+        u64 mtrr_physmask;
+    };
+
+    struct txt_mtrr_state {
+        u64 default_mem_type;
+        u64 mtrr_vcnt;
+        struct txt_mtrr_pair mtrr_pair[TXT_VARIABLE_MTRRS_LENGTH];
+    };
+
+AMD Secure Launch Platforms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+AMD SKINIT Info
+^^^^^^^^^^^^^^^
+
+A placeholder for info specific to AMD SKINIT.
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_amd_info {
+        struct slr_entry_hdr hdr;
+    };
+
+ARM DRTM Environments
+~~~~~~~~~~~~~~~~~~~~~
+
+ARM Info
+^^^^^^^^
+
+A placeholder for info specific to ARM DRTM environments.
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_arm_info {
+        struct slr_entry_hdr hdr;
+    };
+
+UEFI Environments
+~~~~~~~~~~~~~~~~~
+
+To support UEFI bootloaders that may do additional configuration of the Secure
```

Comment:
Hmmm, you are correct that the spec calls it that, but i can't say i have really ever heard it used. Generally I have seen the terms boot manager and boot loader, likely from Rod Smith's early documentation work (https://www.rodsbooks.com/efi-bootloaders/secureboot.html) on all of this.

I guess my question at the end is should we use terms from the UEFI spec that practically no one uses, or use the terminology in common use. Linux documentation opted for the latter.

Reply:
I can ask the same about TCG stuff. Do you think we should agree about the incorrect use or misunderstanding of D-RTM/S-RTM/C-RTM terms?

The everyday use definition needs to be clarified for me. Common use among who? If this spec goes to firmware vendors, I doubt everyday use would be helpful. This gets back to the already asked question: for whom do we write that spec? Let's state that at the beginning. What terminology do we use? Let's say we use common use, not spec definitions.

Reply:
My issue with the acronyms D-RTM and S-RTM is that they are grammatically incorrect. A hyphen in an acronym is meant to denote the combination of two entities for which each has its own acronym, e.g. CD-ROM is Compat Disc Read Only Memory. The correct way to have created the acronym in order to use a hyphen would have been Dynamic Launch Root of Trust for Measurement (DL-RTM).

My issue with the misunderstanding the TCG created with their designation of DRTM/D-RTM vs DCRTM/D-CRTM is what they refer to as the DRTM is really what I would call the Root of Trust for Code Execution (RTCE), ie. the CPU executing externally loaded and mutable code. With this thinking, it quickly becomes much clearer what the trust relationships are and their associated risk exposure. For instance, speculative execution could be considered a violation of the RTCE vs. a ROP gadget in SCRTM code, which I tend to refer to as the true RTM, would be a violation for the RTM.

Now with all of that said, TCG is the governing body over the specific capability we are developing. As such as it is, despite how irritating their poor choice of terminology and grammar is, we should follow it.

Reply:
I see change of DRTM to D-RTM, but what is decision regarding UEFI bootloader vs UEFI OS Loader? I'm just asking for consistence and mutual understanding with reader - I'm not huge fan of UEFI Forum terminology.

Reply:
I have done this,
```
.. note::
   Throughout this specification the term bootloader will be used to
   generically refer to the sofware responsible for loading the OS. This may be
   an x86 bootloader such as GRUB, an embedded system bootloader such as Das
   U-Boot, or an arbitrary UEFI OS Loader.
```